### PR TITLE
gh-112087: Make list_{count, index, contains} to be thread-safe.

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112087

### CODE REVIEW

**Comment by corona10** [comment]
Once the https://github.com/python/cpython/pull/114843 is merged, I would migrate APIs that both atomic operations and `PyList_GetItemRef` is needed.

File: Objects/listobject.c [hunk 1]
```diff
@@ -478,18 +478,21 @@ list_length(PyObject *a)
...
-    Py_ssize_t i;
-    int cmp;
 
-    for (i = 0, cmp = 0 ; cmp == 0 && i < Py_SIZE(a); ++i) {
-        item = PyList_GET_ITEM(a, i);
-        Py_INCREF(item);
-        cmp = PyObject_RichCompareBool(item, el, Py_EQ);
+    for (Py_ssize_t i = 0; ; i++) {
+        PyObject *item = PyList_GetItemRef(aa, i);
+        if (item == NULL && PyErr_ExceptionMatches(PyExc_IndexError)) {
```
Comment (colesbury) [inline Objects/listobject.c]: I think we should have a `static` function that is like `PyList_GetItemRef`, but does not set exceptions and use that here.

File: Objects/listobject.c [hunk 2]
```diff
@@ -272,6 +272,15 @@ PyList_GetItemRef(PyObject *op, Py_ssize_t i)
     return Py_NewRef(PyList_GET_ITEM(op, i));
 }
 
+static inline PyObject*
+_PyList_GetItemRef(PyListObject *op, Py_ssize_t i)
```
Comment (colesbury) [inline Objects/listobject.c]: I think the naming convention is to use the `Py` prefix for global functions (not `static` functions) [^1].

So maybe: `list_get_item_ref` or `list_try_get` or something along those lines.

[^1]: https://peps.python.org/pep-0007/#:~:text=Use%20a%20Py%20prefix%20for%20public%20functions%3B%20never%20for%20static%20functions

**Review by colesbury — COMMENTED** [review]
This looks good to me other than the naming of `_PyList_GetItemRef`.

File: Objects/listobject.c (hunk 2 above)
Comment (corona10) [reply Objects/listobject.c]: Done!

File: Objects/listobject.c [hunk 3]
```diff
@@ -2737,8 +2748,11 @@ list_index_impl(PyListObject *self, PyObject *value, Py_ssize_t start,
             stop = 0;
     }
     for (i = start; i < stop && i < Py_SIZE(self); i++) {
```
Comment (colesbury) [inline Objects/listobject.c]: The `i < Py_SIZE(self)` is probably redundant now.

**Review by colesbury — APPROVED** [review]
LGTM

File: Objects/listobject.c (hunk 3 above)
Comment (corona10) [reply Objects/listobject.c]: Yeah! Nice improvement :)

**Pull request merged**